rustdoc: Allow whitespace as path separator like double colon

GuillaumeGomez opened the pull request:
Fixes https://github.com/rust-lang/rust/issues/108447.

I think it makes sense since it allows more common cases, however it also makes the syntax heavier. Not sure what the rest of the team thinks about it. In any case we'll need to go through FCP.

Full explanation for the changes is available [here](https://github.com/rust-lang/rust/pull/108537#issuecomment-1589480564).

r? @notriddle

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/4285245478/jobs/7463292493) [(plain)](https://github.com/rust-lang/rust/commit/71c779ef9e7f3b0bb75cdd4db2a040220eb8227d/checks/11632634322/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
info: ES-Check: there were no ES version matching errors!  🎉
+ eslint -c ../src/librustdoc/html/static/.eslintrc.js ../src/librustdoc/html/static/js/externs.js ../src/librustdoc/html/static/js/main.js ../src/librustdoc/html/static/js/scrape-examples.js ../src/librustdoc/html/static/js/search.js ../src/librustdoc/html/static/js/settings.js ../src/librustdoc/html/static/js/source-script.js ../src/librustdoc/html/static/js/storage.js

/checkout/src/librustdoc/html/static/js/search.js
  124:14  error  'isWhitespace' is defined but never used. Allowed unused vars must match /^_/u  no-unused-vars
  341:40  error  'prevIsClosingGeneric' is not defined                                           no-undef
  506:48  error  Strings must use doublequote                                                    quotes
  590:48  error  Strings must use doublequote                                                    quotes
  607:19  error  'c' is assigned a value but never used. Allowed unused vars must match /^_/u    no-unused-vars

✖ 5 problems (5 errors, 0 warnings)
  2 errors and 0 warnings potentially fixable with the `--fix` option.
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/4286276423/jobs/7465627119) [(plain)](https://github.com/rust-lang/rust/commit/521fae6d11a6a1590a37cb5d4adf9a48d9563720/checks/11635783092/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Prepare all required actions
Getting action download info
Download action repository 'actions/checkout@v3' (SHA:ac593985615ec2ede58e132d2e21d2b1cbd6127c)
Download action repository 'rust-lang/simpleinfra@master' (SHA:3aac11d64ed0b00fb4926ce853d58cf88de1213a)
Complete job name: PR (x86_64-gnu-tools, false, ubuntu-20.04-xl)
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  CI_JOB_NAME: x86_64-gnu-tools
---
.......... (90/97)
......    (97/97)


/checkout/tests/rustdoc-gui/search-tab-change-title-fn-sig.goml search-tab-change-title-fn-sig... FAILED
[ERROR] (line 64) Error: Evaluation failed: The following errors happened: [`In Names (0)` doesn't start with `In Function Parameters` (for STARTS_WITH check)]: for command `assert-text: ("#search-tabs > button:nth-of-type(1)", "In Function Parameters", STARTS_WITH)`
Build completed unsuccessfully in 0:02:38
```

</details>

GuillaumeGomez commented on src/librustdoc/html/static/js/search.js:
```diff
@@ -590,10 +670,10 @@ function initSearch(rawSearchIndex) {
      * arg = path [generics]
      * arg-without-generic = path
-     * type-sep = COMMA/WS *(COMMA/WS)
+     * type-sep = COMMA *(COMMA)
```
The most recent talk about this was in https://github.com/rust-lang/rust/pull/90630. Otherwise it goes way before 1.0 as far as I see. Quite the old choice.

> Whether we wind up making this change or not, the final decision should be added to https://github.com/rust-lang/rust/blob/master/src/doc/rustdoc/src/how-to-read-rustdoc.md#the-search-interface so that users know what's going on.

Agreed. I'll do it if we agree that we want this change (and that we want it this way).

GuillaumeGomez replied:
Didn't find where the whitespace was introduced so thanks for that!

So I guess with this, we're putting back the original meaning.

bors commented:
:umbrella: The latest upstream changes (presumably #108620) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/4313664827/jobs/7525738451) [(plain)](https://github.com/rust-lang/rust/commit/e747356fd630ec8a8bf07ed122877bd7b4baa514/checks/11714181340/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Prepare all required actions
Getting action download info
Download action repository 'actions/checkout@v3' (SHA:ac593985615ec2ede58e132d2e21d2b1cbd6127c)
Download action repository 'rust-lang/simpleinfra@master' (SHA:0fdabd83e1d3faaa8e9cfd7c00031e3a92997344)
Complete job name: PR (x86_64-gnu-tools, false, ubuntu-20.04-xl)
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  CI_JOB_NAME: x86_64-gnu-tools
---
Successfully built eedd0204933f
Successfully tagged rust-ci:latest
Built container sha256:eedd0204933ff5e0310093718166894ca12063ad4d6a0ef136bf05d18d3db7a2
Uploading finished image to https://ci-caches.rust-lang.org/docker/2022246eea0fceb8dd4e4dc9e56d028016fd8dd73ce83342144a1d278bb5faf6c136af0476a9cbe3f734e6749b99181e6738ca41b0f413c1eb4f076dd7a84292
upload failed: - to s3://rust-lang-ci-sccache2/docker/2022246eea0fceb8dd4e4dc9e56d028016fd8dd73ce83342144a1d278bb5faf6c136af0476a9cbe3f734e6749b99181e6738ca41b0f413c1eb4f076dd7a84292 Unable to locate credentials
[CI_JOB_NAME=x86_64-gnu-tools]
---
.......... (90/97)
......    (97/97)


/checkout/tests/rustdoc-gui/search-tab-change-title-fn-sig.goml search-tab-change-title-fn-sig... FAILED
[ERROR] (line 64) Error: Evaluation failed: The following errors happened: [`In Names (0)` doesn't start with `In Function Parameters` (for STARTS_WITH check)]: for command `assert-text: ("#search-tabs > button:nth-of-type(1)", "In Function Parameters", STARTS_WITH)`
Build completed unsuccessfully in 0:02:38
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #108677) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

GuillaumeGomez committed:
Use whitespace as path separator like double colon

GuillaumeGomez committed:
Update rustdoc-js-std tests

GuillaumeGomez committed:
Add "vec new" test

GuillaumeGomez committed:
Add documentation for paths in rustdoc search

GuillaumeGomez committed:
Update rustdoc-gui test

GuillaumeGomez commented on tests/rustdoc-js-std/vec-new.js:
```diff
@@ -1,9 +1,20 @@
+        ],
+    },
+    {
+        'query': 'Vec::new',
```
... I'm ashamed of myself.

GuillaumeGomez replied:
Fixed that. Not sure what happened.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-14`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/5248179703/jobs/9479316638) [(plain)](https://github.com/rust-lang/rust/commit/68a20ae8d10c4deb5bab874009df789081bcf674/checks/14200848069/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Testing never.js ... OK
Testing option-type-signatures.js ... OK
Testing osstring-to-string.js ... OK
Testing parser-errors.js ... FAILED
[ query `->a<>b`]==> Different values for field `error`:
EXPECTED: `"Expected `,` after `>`, found `b`"`
RESULT:   `"Expected `,` after ,>, found `b`"`
Testing parser-generics.js ... OK
Testing parser-ident.js ... OK
Testing parser-literal.js ... OK
Testing parser-paths.js ... OK
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-14`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/5248773728/jobs/9480708624) [(plain)](https://github.com/rust-lang/rust/commit/eabfbd2171b0087e99bcb9c2ab1439c459780cf0/checks/14202661496/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Testing never.js ... OK
Testing option-type-signatures.js ... OK
Testing osstring-to-string.js ... OK
Testing parser-errors.js ... FAILED
[ query `->a<>b`]==> Different values for field `error`:
EXPECTED: `"Expected `,` after `>`, found `b`"`
RESULT:   `"Expected `,` after ,>, found `b`"`
Testing parser-generics.js ... OK
Testing parser-ident.js ... OK
Testing parser-literal.js ... OK
Testing parser-paths.js ... OK
```

</details>

GuillaumeGomez commented:
I forgot to update all uses of `extra`, which is why the std test failed. Also: the tester is picking an older version of `search.js`. This is quite annoying. Opening an issue for that.

GuillaumeGomez committed:
Correctly display whitespace characters in search error

GuillaumeGomez commented on tests/rustdoc-js-std/parser-generics.js:
```diff
@@ -29,9 +29,9 @@ const PARSED = [
-        userQuery: "p<> u8",
+        userQuery: "p<>,u8",
         error: null,
     },
```
It's definitely a blind spot. Great catch!

GuillaumeGomez committed:
Fix eBNF and handling of whitespace characters when not in a path

GuillaumeGomez commented:
Fixed the corner case you found, updated the eBNF to take it into account and uploaded generated docs [here](https://rustdoc.crud.net/imperio/rustdoc-search-whitespace-as-separator/std/index.html).

GuillaumeGomez commented:
Writing small explanation for other reviewers about this change:

The main motivation for this change is to allow more common cases. When people write `vec new`, I suppose they mean `vec::new` and not `vec,new`. 

The original rationale for using whitespace characters to separate function arguments no longer applies. More details about this thanks to @notriddle archelogy:

> From https://github.com/rust-lang/rust/pull/23289, it looks like the original design for this feature used exclusively commas, not spaces. But since there was no test case for it, the feature was broken by https://github.com/rust-lang/rust/pull/49966 (multi-query search), which also wanted to use the comma.
>
> Since multi-query support was removed (https://github.com/rust-lang/rust/pull/93852), the original reason for supporting this is gone.

So what will change exactly? Whitespace characters will not be considered the same as `,` anymore but as equivalent of `::`. So:

```
query:  Vec new
before: Vec,new
after:  Vec::new
```

Another advantage of this approach is that will make `FnOnce (u32)` and `FnOnce,(u32)` to mean different things.

You can test the feature [here](https://rustdoc.crud.net/imperio/rustdoc-search-whitespace-as-separator/std/index.html).

Manishearth commented:
My main concern was going to be fuzziness: I would like spaces to search fuzzily, but I believe path search already does this (e.g `alloc::iter` does find `alloc::vec::Vec::into_iter()` and friends. So this seems good.

bors commented:
:umbrella: The latest upstream changes (presumably #112609) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

GuillaumeGomez commented:
> My main concern was going to be fuzziness: I would like spaces to search fuzzily, but I believe path search already does this (e.g `alloc::iter` does find `alloc::vec::Vec::into_iter()` and friends. So this seems good.

Just to confirm: yes they do.

GuillaumeGomez commented:
Fixed merge conflict.

camelid commented:
@jsha your checkbox is suddenly checked even though I don't see any history of it being checked. Did you check it and then github lost the edit history, or is this a github bug, and we should uncheck it?

GuillaumeGomez commented:
> My only question is regarding the `FnOnce (u32)` example. How would that be treated with this new change? I'd hope not as `FnOnce::(u32)`. Also, what about `FnOnce u32` (no parentheses)?

`(` and `)` both throw an error if they appear in the search query currently.

GuillaumeGomez commented:
@bors r=notriddle rollup

bors commented:
:pushpin: Commit e4ee3298656b74dd6be810862b5dd4cb8d2825b8 has been approved by `notriddle`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=notriddle e4ee3298656b74dd6be810862b5dd4cb8d2825b8 -->
<!-- homu: {"type":"Approved","sha":"e4ee3298656b74dd6be810862b5dd4cb8d2825b8","approver":"notriddle","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit e4ee3298656b74dd6be810862b5dd4cb8d2825b8 with merge bf0e22b298e282dec3589f70c67e401588d6c693...
<!-- homu: {"type":"BuildStarted","head_sha":"e4ee3298656b74dd6be810862b5dd4cb8d2825b8","merge_sha":"bf0e22b298e282dec3589f70c67e401588d6c693"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/5438059626/jobs/9889991171)
Approved by: notriddle
Pushing bf0e22b298e282dec3589f70c67e401588d6c693 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"notriddle","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/5438059626/jobs/9889991171"},"merge_sha":"bf0e22b298e282dec3589f70c67e401588d6c693"} -->

rust-timer commented:
Finished benchmarking commit (bf0e22b298e282dec3589f70c67e401588d6c693): [comparison URL](https://perf.rust-lang.org/compare.html?start=32d81eccd64513bacef9dfa1574543ada6b45d85&end=bf0e22b298e282dec3589f70c67e401588d6c693&stat=instructions:u).

### Overall result: ❌ regressions - ACTION NEEDED


**Next Steps**: If you can justify the regressions found in this perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please open an issue or create a new PR that fixes the regressions, add a comment linking to the newly created issue or PR, and then add the `perf-regression-triaged` label to this PR.

@rustbot label: +perf-regression
cc @rust-lang/wg-compiler-performance



### [Instruction count](https://perf.rust-lang.org/compare.html?start=32d81eccd64513bacef9dfa1574543ada6b45d85&end=bf0e22b298e282dec3589f70c67e401588d6c693&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.9% | [0.4%, 1.5%] | 25    |
| Regressions ❌ <br /> (secondary)  | 1.0% | [0.4%, 1.5%] | 8     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.9% | [0.4%, 1.5%] | 25    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=32d81eccd64513bacef9dfa1574543ada6b45d85&end=bf0e22b298e282dec3589f70c67e401588d6c693&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 3.3%  | [1.6%, 5.1%]   | 2     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -1.3% | [-1.3%, -1.3%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 1.8%  | [-1.3%, 5.1%]  | 3     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=32d81eccd64513bacef9dfa1574543ada6b45d85&end=bf0e22b298e282dec3589f70c67e401588d6c693&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 3.6%  | [0.5%, 5.6%]   | 8     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.4% | [-0.4%, -0.4%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 3.1%  | [-0.4%, 5.6%]  | 9     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=32d81eccd64513bacef9dfa1574543ada6b45d85&end=bf0e22b298e282dec3589f70c67e401588d6c693&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 661.67s -> 660.214s (-0.22%)

nnethercote commented:
Looks like `incr-full` and `incr-patched` results just started getting noisy for some benchmarks :/

@rustbot label: +perf-regression-triaged
